### ISSUE TITLE
Generic event types lead to "impossible" listener registrations

### ISSUE THREAD

**Issue by Legioth** [vaadin/flow#4085]
The `addListener` method in `ComponentEventNotifier` has the following signature: `<T extends ComponentEvent<?>> Registration addListener(Class<T> eventType, ComponentEventListener<T> listener)`. This causes problems with events for which source type isn't defined, e.g. the current `public class ClickEvent<C extends Component> extends ComponentEvent<C>`.

The basic problem is that with `myComponent.addListener(ClickEvent.class, e -> { ... })`, the `T` parameter will be `ClickEvent`. The Eclipse compiler deals with this by assuming the most generic definition, i.e. `ClickEvent<Component>`, whereas `javac` refuses to compile because the exact type cannot be inferred.

The easiest way of writing something that compiles is to do a raw cast of the listener parameter, e.g. `( ComponentEventListener)  e -> { ... }`. This has the drawback that the type of `e` will be `ComponentEvent`, i.e. no click event methods are available. To actually receive a click event, the type of the first parameter must be e.g. `Class<ClickEvent<TextField>>`, which in turn requires casting through the raw `Class` type, i.e.  `addListener((Class<ClickEvent<TextField>>) (Class) ClickEvent.class, e -> { ... });`.

The easy way out would be to define the source type of events like `ClickEvent` as `Component`. This would mean we cannot create more specific event subclasses, such as `ButtonClickEvent` that would always have `Button` as the source type. This could to some degree be dealt with by defining an `AbstractClickEvent` that doesn't require a specific source type and then extend it as `public class ClickEvent extends AbstractClickEvent<Component>`. The drawback here is that the constructor, including all @EventData definitions for the parameters would still have to be duplicated in each sub class.

Another alternative could be to provide a static helper for producing something that the compiler sees as  `Class<ClickEvent<SomeComponent>>`. This would enable doing something like `myComponent.addListener(ClickEvent.fromComponent(TextField.class), e -> { ... })` which would have `e.getSource()` declared to have the type `TextField`. One problem with this approach is however that it's not very undiscoverable. The other problem is that there's nothing that prevents using `ClickEvent.fromComponent(TextField.class)` with any other component type, thus causing `ClassCastException` when the instance returned from `getSource()` wouldn't have the expected type.

**Comment by pleku** [comment]
>The easy way out would be to define the source type of events like ClickEvent as Component. This would mean we cannot create more specific event subclasses, such as ButtonClickEvent that would always have Button as the source type.

We cannot ? What is blocking from using a specific component subclass as return type for `getSource()` ?

In previous framework versions, you had to declare the events separately for each component if you want to get the source type right. Now we get the most frequent use case (declaring a component specific event) working easier, but it doesn't work for the generic `addListener(...)` case. I'm not sure if we should change back to not have the source type work out of the box for the easy case in order to make events with generic type work via `addListener(...)`.

All components should provide the event-listener API that is most commonly used through their public API, with specific types defined and no compile time errors. When that is not done, it would be nice to get the `addListener(...)` also work. I don't have any good ideas for this. One option might be adding eg. `ComponentClickEvent` etc. where the generic type is always `Component` and thus it would be possible to use it with `addListener`.

Though I can't at the moment come up with any _good_ suggestion on how to handle this... adding "generic"

**Comment by pleku** [comment]
The events that are broken with `addListener` are `ClickEvent`, `FocusEvent` and `BlurEvent`.

Let's investigate first why this works differently in different JVMs, is it a bug or a feature.

One thing to notice also is that the original user wanted get a `ClickEvent` from `TextField`, no idea why focus event was not suitable for this.

**Comment by Legioth** [comment]
I think a more appropriate example for this case is in the scope of layout click listeners, i.e. #2465.

The first fundamental problem is that the type of `ClickEvent.class` is `Class<ClickEvent>`, which also means that the `<T extends ComponentEvent<?>>` type parameter of `addListener` will always be `ClickEvent` and never e.g. `ClickEvent<Div>`.

The second fundamental problem is that an instance method cannot have type parameters that depend on the instance type. This means that there is no way of defining a method called as `instance.method(e -> { ... })` where `e` would be typed based on `instance`. The instance from which to derive the type information must be a method parameter, e.g. `SomeClass.method(instance, e -> { ... })`.

Two variants of the same event type (e.g. `ClickEvent` and `ComponentClickEvent`) will probably only cause confusion about which to use in which cases. Even if using the type that allows the code to compile with `javac`, their `event.getSource()` would still be typed to `Component`, thus negating the main point having the generic types there in the first hand.

Seen from another point of view, the problem is really that the generic `addListener` method is declared as `public`. That method is there to make it possible to create notifier mixins used as e.g. `public class NativeButton extends HtmlContainer implements ClickNotifier<NativeButton>`. The `addListener` method itself isn't really intended for general use, but it's tempting to try to use it in that way since it shows up with auto completion in the IDE.

It might make sense to move the public `addListener` method from `Component` to `ComponentUtil`. "Hiding" the method in this way will decrease the risk of anyone trying to use it in the "wrong" way. There could still be a `protected` method in `Component` with the same signature to be used for components that accept listeners without a notifier mixin interface.

If we instead want to make it possible to add listeners with generic source types, then we'd have to move a way from using the event class literal as the way of identifying the event in the code that the application developer uses for registering a listener. This could be done by moving the registration method to the event class instead: `ClickEvent.listen(myComponent, e -> { ... })`. This would instead require each event type to define an additional static method that would delegate (with an ugly raw casts) to `ComponentUtil.addListener`.

**Comment by gilberto-torrezan** [comment]
Discussed with team and concluded that changing addListener to package protected and adding a method for it to ComponentUtil is fine. Remember to update the docs if necessary https://github.com/vaadin/flow-and-components-documentation/search?l=AsciiDoc&q=addlistener&type=

**Comment by denis-anisimov** [comment]
I'm not sure what should I do with `ComponentEventNotifier`.
It should be removed I suppose.
But then there are subinterfaces which are mix-ins like `AttachNotifier`.
And since they cannot delegate anymore to the `ComponentEventNotifier` method they are not mix-ins anymore and I need to go through all implementors and add the implementation to each of them.

So do I understand right that  `AttachNotifier` and other subinterfaces are not mix-ins anymore ?

**Comment by Legioth** [comment]
> So do I understand right that `AttachNotifier` and other subinterfaces are not mix-ins anymore ?

I'd say that they are still mixins in the sense that the interface doesn't introduce any abstract method. The `default` method should just call `ComponentUtil.addListener((Component) this, ...)` instead of `this.addListener(...)`.

**Comment by denis-anisimov** [comment]
Hmm......
`(Component)this` ?
So in reality we assume that those interfaces are always extended by component.
Is it safe?
I'm not sure.

**Comment by Legioth** [comment]
It's not safe in all cases, but I think it's safe enough in practice. We could maybe slightly reduce the risk of someone accidentally adding the mixin to some non-component class by making the mixin extend something that components typically inherit, e.g. `HasElement`. That's not foolproof either, but should reduce the risk of accidents.

**Comment by denis-anisimov** [comment]
Since it's a mixin with the default impl for the method I would say we should throw if it doesn't extend `Component` so in this case the developer should override and implement this method on its own.

**Comment by Legioth** [comment]
`(Component) this` will throw automatically, though a more informative message might of course also be beneficial.

**Comment by denis-anisimov** [comment]
Yeah ,I'm talking about message like "the default impl doesn't know how to add a listener, please implements the method by yourself".
